feature/s3/transfermanager/putobject.go

```diff
@@ -0,0 +1,772 @@
+
+	// The base64-encoded, 256-bit SHA-256 digest of the object.
+	ChecksumSHA256 *string
+
+	// Entity tag for the uploaded object.
+	ETag *string
+
+	// If the object expiration is configured, this will contain the expiration date
+	// (expiry-date) and rule ID (rule-id). The value of rule-id is URL encoded.
+	Expiration *string
+
+	// The bucket where the newly created object is put
+	Bucket *string
+
+	// The object key of the newly created object.
+	Key *string
+
+	// If present, indicates that the requester was successfully charged for the
+	// request.
+	RequestCharged types.RequestCharged
+
+	// If present, specifies the ID of the Amazon Web Services Key Management Service
+	// (Amazon Web Services KMS) symmetric customer managed customer master key (CMK)
+	// that was used for the object.
+	SSEKMSKeyId *string
+
+	// If you specified server-side encryption either with an Amazon S3-managed
+	// encryption key or an Amazon Web Services KMS customer master key (CMK) in your
+	// initiate multipart upload request, the response includes this header. It
+	// confirms the encryption algorithm that Amazon S3 used to encrypt the object.
+	ServerSideEncryption types.ServerSideEncryption
+
+	// The version of the object that was uploaded. Will only be populated if
+	// the S3 Bucket is versioned. If the bucket is not versioned this field
+	// will not be set.
+	VersionID *string
+}
+
+func (c Client) Upload(ctx context.Context, input *PutObjectInput, opts ...func(*Options)) (*PutObjectOutput, error) {
+	i := uploader{in: input, cfg: c, ctx: ctx}
+	// Copy ClientOptions
+	clientOptions := make([]func(*s3.Options), 0, len(c.options.PutClientOptions)+1)
+	clientOptions = append(clientOptions, func(o *s3.Options) {
+		o.APIOptions = append(o.APIOptions,
+			middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+			addFeatureUserAgent, // yes, there are two of these
+			func(s *smithymiddleware.Stack) error {
+				return s.Finalize.Insert(&setS3ExpressDefaultChecksum{}, "ResolveEndpointV2", smithymiddleware.After)
+			},
+		)
+	})
+	clientOptions = append(clientOptions, c.options.PutClientOptions...)
+	i.cfg.options.PutClientOptions = clientOptions
+	for _, opt := range opts {
+		opt(&i.cfg.options)
+	}
+
+	return i.upload()
+}
+
+type uploader struct {
+	ctx context.Context
+	cfg Client
+
+	in *PutObjectInput
+
+	//totalSize int64 // set to -1 if the size is not known
+}
+
+func (u *uploader) upload() (*PutObjectOutput, error) {
+	if err := u.init(); err != nil {
+		return nil, fmt.Errorf("unable to initialize upload: %w", err)
+	}
+	defer u.cfg.options.partPool.Close()
+
+	r, _, cleanUp, err := u.nextReader()
+
+	if err == io.EOF {
+		return u.singleUpload(r, cleanUp)
+	} else if err != nil {
+		cleanUp()
+		return nil, err
+	}
+
+	mu := multiUploader{
+		uploader: u,
+	}
+	return mu.upload(r, cleanUp)
+}
+
+func (u *uploader) init() error {
+	if err := validateSupportedARNType(aws.ToString(u.in.Bucket)); err != nil {
```

Comment:
Was a bit curious about this to spend some time looking into it. S3 Object Lambda is a feature that allows you to front operation on a bucket like `GetObject` with a Lambda that can do transformations on the object. When you create an Object Lambda, you get an `ARN` that you call instead of the ARN of the bucket, since it's a new access point. This access point only supports GET, HEAD and LIST operations, so a PutObject call will fail (I suppose, haven't tried it to see *how* exactly it fails)

IMO this seems a bit niche to add a validation transfer manager